src/ucp/rndv/proto_rndv.c

```diff
@@ -48,17 +53,28 @@ ucp_proto_rndv_ctrl_reg_md_map(const ucp_proto_rndv_ctrl_init_params_t *params)
         /* Check the memory domain requires remote key, and capable of
          * registering the memory type
          */
-        md_index = ucp_proto_common_get_md_index(&params->super.super, lane);
-        md_attr  = &worker->context->tl_mds[md_index].attr;
+        ep_sys_dev = ucp_proto_common_get_sys_dev(&params->super.super, lane);
+        md_index   = ucp_proto_common_get_md_index(&params->super.super, lane);
+        md_attr    = &worker->context->tl_mds[md_index].attr;
         if (!(md_attr->cap.flags & UCT_MD_FLAG_NEED_RKEY) ||
             !(md_attr->cap.reg_mem_types & UCS_BIT(params->mem_info.type))) {
             continue;
         }
 
-        reg_md_map |= UCS_BIT(md_index);
-    }
+        *md_map |= UCS_BIT(md_index);
 
-    return reg_md_map;
+        if (ep_sys_dev < 64) {
+            *sys_dev_map |= UCS_BIT(ep_sys_dev);
+
+            mem_sys_dev = params->super.super.select_param->sys_dev;
+            status      = ucs_topo_get_distance(mem_sys_dev, ep_sys_dev,
+                                                sys_distance);
```

Comment:
Is there an assumption here that sys_device indices are symmetric across processes? (i.e if we sys_dev0 = mlx5_0, sys_dev1 = mlx5_1, sys_dev2 = gpu0, sys_dev3 = gpu1, then this is true for any UCX process in the job). If so, then user setting CUDA_VISIBLE_DEVICES at different sites (a pretty common scenario) would break the assumption because we populate sys devices dynamically in different UCTs (at least in cuda transport). 

Also, is there an expectation that if remote sys_device and local system_device are compared for distance then nvlink reachability (and associated bw/latency) is considered? If so, then we would need remote system device to be locally viewable (not always true because CUDA_VISIBLE_DEVICES can limit the view) and that device have unique system device id that allows for cuda-ipc reachability and bandwidth/latency estimates based on that. Maybe this is where remote system device translation becomes necessary.

Reply:
1. there is no assumption that sys device indices are symmetric, that's why rkey contains distance values (packed into 8-bit floating-point value)
2. remote and local system devices are never compared to each other. we only compare the distance between a system device (mem_sys_dev above) and a communications device (ep_sys_dev) on the same machine. nvnlink does not fit well to this model currently since once communications device represents all GPUs on the system